services/traction/api/db/models/v1/tenant_job.py

```diff
@@ -0,0 +1,107 @@
+that need to happen by the innkeeper and the tenant before issuance is possible.
+
+"""
+import uuid
+from enum import Enum
+
+from sqlmodel import Field
+from sqlalchemy import Column, text, UniqueConstraint, select
+from sqlalchemy.dialects.postgresql import UUID, JSON
+from sqlmodel.ext.asyncio.session import AsyncSession
+
+from api.db.models.base import (
+    TimestampModel,
+    TenantScopedModel,
+    StatefulModel,
+)
+
+
+class TenantJobType(str, Enum):
+    endorser = "Endorser"
+    public_did = "Public DID"
+    issuer = "Issuer"
+
+
+class TenantJobStatusType(str, Enum):
+    requested = "Requested"  # requested by tenant
+    approved = "Approved"  # approved by innkeeper, tenant can proceed
+    denied = "Denied"  # denied by innkeeper, job terminates
+    processing = "Processing"  # job is in flight, check state field for progress
+    error = "Error"  # error during processing
+    completed = "Completed"  # job is completed, value in completed_value
+    active = "Active"  # job complete and anything dependent on this can proceed
+    default = "N/A"  # job has not been started.
+
+
+class TenantJobStateType(str, Enum):
+    """
+    state should be different types for different jobs...
+    """
+
+    default = "N/A"  # flow has not been started.
+
+
+class TenantJob(TenantScopedModel, StatefulModel, TimestampModel, table=True):
+    __tablename__ = "tenant_job"
+    __table_args__ = (UniqueConstraint("tenant_id", "job_type"),)
+
+    tenant_job_id: uuid.UUID = Field(
+        sa_column=Column(
+            UUID(as_uuid=True),
+            primary_key=True,
+            server_default=text("gen_random_uuid()"),
+        )
+    )
+
+    job_type: str = Field(nullable=False)
+    data: dict = Field(default={}, sa_column=Column(JSON))
+    comment: str = Field(nullable=True)
+
+    @classmethod
+    async def get_for_tenant(
+        cls: "TenantJob",
+        db: AsyncSession,
+        tenant_id: uuid.UUID,
+        job_type: str,
+    ) -> "TenantJob":
+        """Get TenantJob by tenant id and flow type.
+
+        Find and return the database record.
+        If one does not exist, insert a default and return it.
+
+        Args:
+          db: database session
+          tenant_id: Traction ID of tenant this flow belongs to
+          job_type: the particular job we are interested n
+
+        Returns: The Traction TenantJob (db) record
+
+        Raises:
+
+        """
+
+        q = (
+            select(cls)
+            .where(cls.tenant_id == tenant_id)
+            .where(cls.job_type == job_type)
+        )
+        q_result = await db.execute(q)
+        db_rec = q_result.scalar_one_or_none()
+        if not db_rec:
+            db_rec = cls(
+                tenant_id=tenant_id,
+                job_type=job_type,
+                status=TenantJobStatusType.default,
+                state=TenantJobStateType.default,
+            )
```

Comment:
Why do we want this behaviour, if you GET something that doesn't exist, shouldn't this method error (the caller can catch it if needed)?

Reply:
it's the same thing as configuration or permissions, you want it to exist and it creates the default state. it's not the same as a business data record.

Reply:
This seems like unexpected behaviour, also, is job type unique (a tenant can only run a job once, ever)? maybe rename the method to `get_or_create_tenant_job`?

Reply:
i think if you read the code you will see that it is unique per tenant per job type.

Reply:
That is true for the on job we have, but seems very likely this will be used to things that can be run multiple times.

Reply:
we have 3 jobs, and you are quite right... it is possible in the future that jobs or whatever we end up calling them may be run multiple times. in that case we update the code. or we consider it as we go over all the protocol/tasks/jobs that we have now and are re-factoring.

Reply:
Given this ticket only changes the implementation and does not change the behaviour. We should really be deliberate in the solutions we are building. If 'jobs' are the going to compose other procedures into more useful business level actions, then we should design it to do that form the ground up.

Jobs should be unique based on job type. and function names should be clear in what they do.

Reply:
i'm not sure what you mean? there are 3 jobs: get an endorser connection, get a public did, become an issuer. becoming and issuer depends on the other 2 jobs. so they are unique on job type. each job has value on its own, but some jobs require one (or more) to be completed before they can do their work. i mean, we talked about it at length, we had no way to do a dependency chain. if you are now saying to amalgamate those 3 jobs into one, that's fine, but it's not what we talked about. the true goal of this ticket was to allow us to completely sever the v0 code base. the endorser/publicdid/make issuer was the last process that needed to be extracted. if you want it in one chunk of code that is only for make-issuer, then that's easy.

Reply:
this probably isn't the forum to have this kind of discussion, we have discord. so I say move it there, then we can figure out what to do with code.

Reply:
We have an understanding that jobs is not a final solution and that re-evaluating the purpose and design of tasks, protocols, jobs, workflows, all need a cohesive redesign. To be collaborated on in the future.